### PR TITLE
fix str mutating through a ptr derived from &self

### PR DESCRIPTION
**Pull request by RalfJung**
Found by Miri: In `get_unchecked_mut` (also used by the checked variants internally) uses `str::as_ptr` to create a mutable reference, but `as_ptr` takes `&self`.  This means the mutable references we return here got created from a shared reference, which violates the shared-references-are-read-only discipline!

For this by using a newly introduced `as_mut_ptr` instead.

### CODE REVIEW

File: src/libcore/str/mod.rs [hunk 1]
```diff
@@ -1757,9 +1757,9 @@ mod traits {
         }
         #[inline]
         unsafe fn get_unchecked_mut(self, slice: &mut str) -> &mut Self::Output {
-            let ptr = slice.as_ptr().add(self.start);
+            let ptr = slice.as_mut_ptr().add(self.start);
             let len = self.end - self.start;
-            super::from_utf8_unchecked_mut(slice::from_raw_parts_mut(ptr as *mut u8, len))
+            super::from_utf8_unchecked_mut(slice::from_raw_parts_mut(ptr, len))
```
Comment (RalfJung) [inline src/libcore/str/mod.rs]: Notice that this raw ptr cast here is the "canary" that gave away that the old code was wrong -- it was actually casting `*const u8` to `*mut u8`, which should not have been necessary.

File: src/libcore/str/mod.rs [hunk 2]
```diff
@@ -2213,6 +2213,22 @@ impl str {
...
+    ///
+    /// As string slices are a slice of bytes, the raw pointer points to a
+    /// [`u8`]. This pointer will be pointing to the first byte of the string
+    /// slice.
+    ///
+    /// It is your responsibility to make sure that the string slice only gets
+    /// modified in a way that it remains valid UTF-8.
+    ///
+    /// [`u8`]: primitive.u8.html
+    #[unstable(feature = "str_as_mut_ptr", issue = "0")]
```
Comment (SimonSapin) [inline src/libcore/str/mod.rs]: r+ with a tracking issue

**Review by SimonSapin — APPROVED** [review]

File: src/libcore/str/mod.rs (hunk 2 above)
Comment (RalfJung) [reply src/libcore/str/mod.rs]: I opened a tracking issue in https://github.com/rust-lang/rust/issues/58215, please check if that looks all right.

**Comment by RalfJung** [comment]
Turns out there was another bad use of `as_ptr`, in `str::split_at_mut`. I fixed that as well and reviewed the remaining uses in `str/mod.rs`.

@SimonSapin could you review?

File: src/libcore/str/mod.rs (hunk 2 above)
Comment (hamdahmdalzhrany-ux) [reply src/libcore/str/mod.rs]: > I opened a tracking issue in https://github.com/rust-lang/rust/issues/58215, please check if that looks all right.

**Pull request merged**